src/olympia/files/helpers.py

```diff
@@ -78,34 +108,73 @@ def __init__(self, file_obj):
     def __str__(self):
         return str(self.file.id)
 
-    def _extraction_cache_key(self):
-        return ('%s:file-viewer:extraction-in-progress:%s' %
-                (settings.CACHE_PREFIX, self.file.id))
+    def _cache_key(self, key=None):
+        assert key is not None
+        return '{0}:file-viewer:{1}:{2}'.format(
+            settings.CACHE_PREFIX, key, self.file.id
+        )
+
+    @contextlib.contextmanager
+    def lock(self):
+        """Lock the file-viewer.
+
+        If locked, no one is allowed to run any extraction in `self.src` to
+        avoid files / directories get messed up.
+        """
+        cache_key = self._cache_key(LOCKED)
+        msg = Message(cache_key)
+
+        if msg.get():
+            # Already locked
+            yield True
+        else:
+            # Not yet locked, save flag and delete on exit.
+            msg.save(True, time=LOCKED_TIMEOUT)
```

Comment:
It looks like memcached has a `cas` command for "check and set", it will only set the value if it is the same as when you last read it. Guido writes about it [1] and it looks like you could replace the `set()` call with `cas()` and `get()` with `gets()` if our helpers support it and handle the "it was changed" case as it already being locked.

[1] http://neopythonic.blogspot.ca/2011/08/compare-and-set-in-memcache.html